Google.Api.Generator/Generation/MetadataGenerator.cs

```diff
@@ -0,0 +1,84 @@
+﻿// Copyright 2021 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Google.Gapic.Metadata;
+using static Google.Gapic.Metadata.GapicMetadata.Types;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Google.Api.Generator.Generation
+{
+    internal static class MetadataGenerator
+    {
+        private const string GapicMetadataSchemaVersion = "1.0";
+        private const string GapicMetadataFileDescription = "This file maps proto services/RPCs to the corresponding library clients/methods";
+        private const string GapicMetadataLanguage = "csharp";
+
+        private const string TransportKeyGrpc = "grpc";
+
+        internal static string GenerateGapicMetadataJson(List<ServiceDetails> allServiceDetails)
+        {
+            var gapicMetadata = new GapicMetadata
+            {
+                Schema = GapicMetadataSchemaVersion,
+                Comment = GapicMetadataFileDescription,
+                Language = GapicMetadataLanguage,
+                ProtoPackage = allServiceDetails.First().ProtoPackage,
+                LibraryPackage = allServiceDetails.First().Namespace,
+                Services = {
+                allServiceDetails.ToDictionary(
+                    serviceDetails => serviceDetails.ServiceName,
+                    ServiceForTransportMetadata
+                )}
+            };
+
+            JObject obj = JsonConvert.DeserializeObject<JObject>(
+                gapicMetadata.ToString(),
+                new JsonSerializerSettings { DateParseHandling = DateParseHandling.None }
+            );
+
+            if (obj == null)
+            {
+                throw new InvalidOperationException("Json conversion failed for the Gapic Metadata object.");
+            }
+
+            return obj.ToString(Formatting.Indented) + "\n"; //trailing file newline to please github;

```

Comment:
```suggestion
            return JsonConvert.DeserializeObject<JObject>(
                gapicMetadata.ToString(),
                new JsonSerializerSettings { DateParseHandling = DateParseHandling.None })
                ?.ToString(Formatting.Indented) ?? 
            throw new InvalidOperationException("Json conversion failed for the Gapic Metadata object.");
```

Or even everything in one expresion if you build the gapicMetadata instance inline the above suggestion. Then you can turn the method to an expression bodied one.

Reply:
I would prefer not to do this since in my opinion that would cross the line where terseness stops being useful and becomes a distraction instead.

Reply:
I agree that doing a one expression by creating the gapicMetada instance inline is too much. But I personally do like not having to ask about null for throwing the exception.
But it's also fine if you don't :).